### PR TITLE
ICS28: CCV draft w/ complete unbonding

### PR DESCRIPTION
**Pull request by mpoke**
The partial specification of Cross-Chain Validation. 

Changes to base (i.e., [ccv](https://github.com/cosmos/ibc/tree/marius/ccv-init-and-vsc)):
- The staking hooks that enable the timely completion of unbonding operations (e.g., undelegations) on the provider chain (i.e., waiting for the operations to mature on the consumer chain).

Features not included (to be added to other PRs):
- the evidence / slashing subprotocol that enables the provider chain to punish validators for misbehaviors on the consumer chain;
- the distribution subprotocol that enables the validators to receive rewards for validating on the consumer chain.

## Prerequisites 

- [x] Handle issue https://github.com/cosmos/interchain-security/issues/33 before merging

### CODE REVIEW

File: spec/app/ics-028-cross-chain-validation/technical_specification.md [hunk 1]
```diff
@@ -931,18 +958,78 @@ function onAcknowledgeVSCPacket(packet: Packet) {
...
+      ops.Append(op)
+    }
+    return ops
+}
+```
+- **Initiator:** 
+  - The provider CCV module when receiving an acknowledgement for a `VSCPacket`.
+- **Expected precondition:**
+  - None. 
+- **Expected postcondition:**
```
Comment (josef-widder) [inline spec/app/ics-028-cross-chain-validation/technical_specification.md:1027]: I wonder whether the pseudo code provides additional information that is not present in the expected postconditions? In general, is there is no algorithmic ideas, I have tried to reduce pseudo code as much as possible from specs. But perhaps that is against interchain standards?

**Review by josef-widder — APPROVED** [review]
Looks good to me. The only question is whether pseudo code and the bullet points below the functions may not be redundant. We can discuss this.

**Review by AdityaSripal — APPROVED** [review]
Nice work!

**Comment by AdityaSripal** [comment]
I would hold off on merging until the init spec is approved so it is easier to review

**Comment by mpoke** [comment]
- [ ] Handle issue https://github.com/cosmos/interchain-security/issues/33 before merging

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 2]
```diff
@@ -161,7 +169,41 @@ In the context of VSCs, the CCV module enables the following functionalities:
...
+For a more detailed description of Validator Set Update, take a look at the [technical specification](./technical_specification.md#validator-set-update).
 
 ![Validator Set Update Overview](./figures/ccv-vsc-overview.png?raw=true)
 
-For a more detailed description of Validator Set Update, take a look at the [technical specification](./technical_specification.md#validator-set-update).
\ No newline at end of file
+### Completion of Unbonding Operations
+
+In the context of single-chain validation, the completion of any unbonding operation requires the `UnbondingPeriod` to elapse since the operations was initiated (i.e., the operation MUST reach maturity). 
+In the context of CCV, the completion MUST require also the unbonding operation to reach maturity on all consumer chains (for the [Security Model](#security-model) to be preserved).
```
Comment (josef-widder) [inline spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:179]:
```suggestion
In the context of CCV, the completion MUST require also the unbonding operation to reach maturity on **all** consumer chains (for the [Security Model](#security-model) to be preserved). 
```

**Review by josef-widder — APPROVED** [review]

File: spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md [hunk 3]
```diff
@@ -95,9 +96,16 @@ which MUST NOT complete before reaching maturity on both the provider and all th
...
 Thus, a *VSC reaching maturity* on a consumer chain means that all the unbonding operations that resulted in validator updates included in that VSC have matured on the consumer chain.
 
-> **Background**: An *unbonding operation* is any operation of unbonding an amount of the tokens a validator bonded. Note that the bonded tokens correspond to the validator's voting power. Unbonding operations have two components: 
+> **Background**: An *unbonding operation* is any operation of unbonding an amount of the tokens a validator bonded. Note that the bonded tokens correspond to the validator's voting power. We distinguish between three types of unbonding operations:
+> - *undelegation* - a delegator unbonds tokens it previously delegated to a validator;
+> - *redelegation* - a delegator instantly redelegates tokens from a source validator to a different validator (the destination validator);
+> - *validator unbonding* - a validator is removed from the validator set; note that although validator unbondings do not entail unbonding tokens, they behave similarly to other unbonding operations.
```
[the remark this answers is not in the document]
Comment (mpoke) [reply spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:102]: It's from the SDK documentation, see https://docs.cosmos.network/v0.44/modules/staking/05_end_block.html#unbonding-validators

**Pull request merged**